### PR TITLE
LPS-182893 First steps to add new AICreator plugin to CKEditor

### PR DESCRIPTION
**Pull request by p2kmgcl**
Hey there! As I have discussed with @dsanz in Slack, we have a new [epic in echo team to integrate OpenAI with CKEditor](https://issues.liferay.com/browse/LPS-179483). Currently it is going to be enabled just for web content editing, but we want to contribute this as a regular plugin to be sure we can enable it in more parts of portal easily.

This PR is just a non-final, non-working approach to check if we are aligned with the file structure and the components that we need to create. Everything is hidden behind the `LPS-179483` feature flag.

These are some points I would like to discuss:

1. I have added an empty CKEditor plugin in `aicreator/plugin.js`, which registers a new button that does nothing. I am not sure how to connect this with the rest of the functionality, as it is a React application that needs to be mounted in the DOM. At the moment the modal could be opened with `Liferay.fire()`, but I also need the plugin button to be part of the React application: it has more logic that needs to be added (check [the design documentation](https://www.figma.com/file/1vRXQ8im36A9TUdC66slE1/LPS-179483---Add-integration-to-OpenAI?type=design&node-id=64-201809&t=JJ02iPs1ukiSFOhK-0) for more info).
2. To use the button in `journal-web`, I am using a custom `toolbar` configuration for CKEditor, which is just a temporary solution to quickly include the new button in the toolbar. I still need to have a conversation with Carol (the designer working on this epic) to check how we should integrate this button with some exiting toolbar configuration. @dsanz gave me some documentation for this and I am going to start working on that today.
3. There is some pending backend work to make the actual communication with OpenAI. All communication will pass through the backend and I still need to figure out how we will obtain the corresponding URLs (cc @lfbesada).

Thanks so much for having a look! :smile:

### CODE REVIEW

**Comment by p2kmgcl** [comment]
I just moved the Modal code to the new `ai-creator` module. No more changes :sweat_smile: 
We still need to think how to communicate these two modules, probably exporting a function instead of relying on a global `Liferay.fire` event would be a better idea.

**Comment by p2kmgcl** [comment]
I have been thinking about how we should integrate the editor plugin with OpenAI, and I came up with this idea:

- Add the plugin in CKEditorConfigContributor.java, in the module that will be using the plugin:

```java
String extraPlugins = jsonObject.getString("extraPlugins");

if (Validator.isNotNull(extraPlugins)) {
	extraPlugins = extraPlugins + ",aicreator";
}
else {
	extraPlugins = "aicreator";
}

jsonObject.put("extraPlugins", extraPlugins);
```

- Load the plugin resources in `ckeditor_classic.jsp` (the `_checkIfOpenAIPluginIsEnabled` could be implemented here, or we could contribute it from the `open-ai` module):

```jsp
<% if (_checkIfOpenAIPluginIsEnabled(editorConfigJSONObject) { %>
	<ai-creator:openai-ckeditor-plugin />
<% } %>
```

- Get the Liferay component that is loaded in the JSP from the editor plugin (plugins/aicreator/plugin.js):

```js
(function () {
	const pluginName = 'aicreator';

	if (CKEDITOR.plugins.get(pluginName)) {
		return;
	}

	CKEDITOR.plugins.add(pluginName, {
		init(editor) {
			const button = editor.ui.addButton('AICreator', {
				command: '',
				icon: '',
				label: '',
			});
			
			Liferay.componentReady(
				'ai-creator-openai-ckeditor-plugin'
			).then((plugin) => {
				plugin.mountCKEditorButton(editor, button);
			});
		},
	});
})();
```

This way we can decouple almost everything from CKEditor module. What do you think about that?

**Comment by markocikos** [comment]
@p2kmgcl There may be a simpler way to integrate this, but I'm not sure sure I fully understand all the requirements:
- The button in toolbar of CKEditor should not be a React component. You should use `editor.ui.addButton` to add it. The alternative it to add support for React to base CKEditor, or do some ugliness like direct DOM manipulation on a rendered React component. I am not sure if that would even work, and it could very easily open a ton of issues. Looking at your mockups, there doesn't seem to be anything special about the button, besides the tootip. You should be able to implement that easily without React.
- A click on the button can open a modal with `Liferay.Util.openModal`. You can pass in `url` to that modal, it will open it in an `iframe`. That `url` can be to a JSP, containing a `<react:component>` tag, wired to the React component itself. For example, we are [doing something similar with 'imageselector'](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/itemselector/plugin.js#L354-L370). In the place of the linked code, you have a point of wiring, with modal and editor APIs. The url is taken from config, set in a [ConfigContributor](https://github.com/liferay/liferay-portal/blob/master/modules/apps/adaptive-media/adaptive-media-journal-editor-configuration/src/main/java/com/liferay/adaptive/media/journal/editor/configuration/internal/AMJournalEditorConfigContributor.java#L127).

Would this work for you?

**Comment by p2kmgcl** [comment]
Hey @markocikos!

Sorry for the delay, I was attending DevCon and didn't have time to answer here. I think that should work perfectly. I was trying to decouple the OpenAI code from CKEditor, but since this is going to be a plugin, I agree that we should create the button and open the modal from the plugin itself.

I am refactoring this to make it happen.

Thanks!

**Comment by p2kmgcl** [comment]
@markocikos done! This should be ready to go

I have exluded the modal creation and just added a new CKEditor plugin which opens a modal, as you suggested :smile:

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/aicreator/plugin.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	if (CKEDITOR.plugins.get(pluginName)) {
+		return;
+	}
+
+	CKEDITOR.plugins.add(pluginName, {
+		init(editor) {
+			const plugin = this;
+
+			const handleDialogClose = () => {
+				editor.insertText('Sample text');
```
[the remark this answers is not in the document]
Comment (p2kmgcl) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/aicreator/plugin.js:27]: Yep, I will take the real content from the close event, but I still need to build the modal JSP to do it :sweat_smile: 

There will be a follow up

**Comment by ealonso** [comment]
Hey @dsanz,
Could you please move this pr? it is blocking us the Q2, we need to move it today, if not we are not going to be able to finish it on time.

**Comment by dsanz** [comment]
Hey @ealonso we can forward it, yes, but @p2kmgcl said he was preparing a follow up.

What's the course of action then? Should we wait for the followup?

**Comment by dsanz** [comment]
ci:forward

**Comment by ealonso** [comment]
ci:forward:force

**Pull request closed without merging**